src/dbnode/persist/fs/retriever.go

```diff
@@ -321,28 +325,34 @@ func (r *blockRetriever) fetchBatch(
 				onRetrieveSeg = ts.NewSegment(dataCopy, nil, ts.FinalizeHead)
 				dataCopy.AppendAll(data.Bytes())
 			}
-			if tags := req.indexEntry.EncodedTags; len(tags) != 0 {
-				tagsCopy := r.bytesPool.Get(len(tags))
-				tagsCopy.IncRef()
-				tagsCopy.AppendAll(req.indexEntry.EncodedTags)
-				tagsCopy.DecRef()
+			if tags := req.indexEntry.EncodedTags; tags != nil {
 				decoder := tagDecoderPool.Get()
-				decoder.Reset(tagsCopy)
+				// DecRef because we're transferring ownership from the index entry to
+				// the tagDecoder which will IncRef().
+				tags.DecRef()
+				decoder.Reset(tags)
 				req.tags = decoder
 			}
+		} else {
+			// If we didn't transfer ownership of the tags to the decoder above, then we
+			// no longer need them and we can can finalize them.
+			if tags := req.indexEntry.EncodedTags; tags != nil {
+				tags.DecRef()
+				tags.Finalize()
+			}
```

Comment:
Regardless of whether we call on retrieve or not, since we're now pooling `req.indexEntry.ID` should we always also `req.indexEntry.ID.Finalize()`?